src/ddmd/expressionsem.d

```diff
@@ -8771,6 +8784,101 @@ Expression binSemanticProp(BinExp e, Scope* sc)
     return null;
 }
 
+/********************************************************************************
+ * Helper function to resolve `@property` functions in a `BinAssignExp`.
+ * It rewrites expressions of the form, for example, `s.x += 1` to `s.x(s.x() += 1)`
+ * if `s` has a both a getter and setter `@property` named `x`.
+ * Params:
+ *      e = the expression to rewrite
+ *      sc = the semantic scope
+ * Returns:
+ *      the rewritten expression if the procedure succeeds, an `ErrorExp` if the
+ *      and error is encountered, or `null` if `e.e1` is not an `@property` function.
+ */
+Expression binSemanticProp(BinAssignExp e, Scope* sc)
+{
+    if (Expression ex = binSemantic(e, sc))
+        return ex;
+    Expression e1x = resolveProperties(sc, e.e1);
+    Expression e2x = resolveProperties(sc, e.e2);
+    if (e1x.op == TOKerror)
+        return e1x;
+    if (e2x.op == TOKerror)
+        return e2x;
+
+    //********************************************************************
+    // Check for property assignment.
+    // https://issues.dlang.org/show_bug.cgi?id=8006
+    if (e1x.op == TOKcall)
+    {
+        CallExp ce = cast(CallExp)e1x;
+
+        // Consider s.x() += 1;
+        // We want to rewrite that expression to s.x(s.x() + 1);
+
+        // Only rewrite @property functions.  Don't rewrite @ref @property functions.
```

Comment:
First of all, this already works today:
```
struct TInt
{
    int mX;

    ref @property int x()
    {
        return mX;
    }

    alias x this;
}

void main()
{
    TInt t;

    t.x += 4;
    assert(t.mX == 4);
    t.x -= 2;
    assert(t.mX == 2);
    t.x *= 4;
    assert(t.mX == 8);
    t.x /= 2;
    assert(t.mX == 4);
    t.x %= 3;
    assert(t.mX == 1);
    t.x <<= 3;
    assert(t.mX == 8);
    t.x >>= 1;
    assert(t.mX == 4);
    t.x >>>= 1;
    assert(t.mX == 2);
    t.x &= 0xF;
    assert(t.mX == 0x2);
    t.x |= 0x8;
    assert(t.mX == 0xA);
    t.x ^= 0xF;
    assert(t.mX == 0x5);
    t.x ^^= 2;
    assert(t.mX == 25);

    // same as test above, but through the `alias this`
    t = 0;
    t += 4;
    assert(t.mX == 4);
    t -= 2;
    assert(t.mX == 2);
    t *= 4;
    assert(t.mX == 8);
    t /= 2;
    assert(t.mX == 4);
    t %= 3;
    assert(t.mX == 1);
    t <<= 3;
    assert(t.mX == 8);
    t >>= 1;
    assert(t.mX == 4);
    t >>>= 1;
    assert(t.mX == 2);
    t &= 0xF;
    assert(t.mX == 0x2);
    t |= 0x8;
    assert(t.mX == 0xA);
    t ^= 0xF;
    assert(t.mX == 0x5);
    t ^^= 2;
    assert(t.mX == 25);
}
```
But more importantly, we found that we broke code in DUB when not excluding `ref @property` functions.  The specific idiom that broke was added to the test suite [here](https://github.com/JinShil/dmd/blob/ae7ddcadfb0c25f2a1afcec041c6bb9a7aef2522/test/runnable/test8006.d#L354-L372).

So it `ref @property` functions were excluded in an effort to be conservative with this change, and avoid code breakage.